### PR TITLE
[quant] Add serialization for nn.quantized.Linear module

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #21930 nn.quantized.Relu and nn.quantize.Quantize/DeQuantize modules&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15554224/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#21925 [quant] Add serialization for nn.quantized.Linear module**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15483071/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #21921 Implementation of nn.quantized.linear module&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D15375695/)

att

Differential Revision: [D15483071](https://our.internmc.facebook.com/intern/diff/D15483071/)

### CODE REVIEW

File: torch/nn/quantized/modules/linear.py [hunk 1]
```diff
@@ -42,3 +50,71 @@ def __init__(self, in_features, out_features, bias=True):
     def forward(self, x):
         Y_q = torch.ops.quantized.fbgemm_linear(x, self._packed_weight, self.bias, self.output_scale, self.output_zero_point)
         return Y_q
+
+    def state_dict(self, destination=None, prefix='', keep_vars=False):
```
Comment (dzhulgakov) [inline torch/nn/quantized/modules/linear.py]: https://github.com/pytorch/pytorch/pull/21933 might be helpful

File: torch/serialization.py [hunk 2]
```diff
@@ -294,8 +294,8 @@ def persistent_id(obj):
     pickler = pickle_module.Pickler(f, protocol=pickle_protocol)
     pickler.persistent_id = persistent_id
     pickler.dump(obj)
-
     serialized_storage_keys = sorted(serialized_storages.keys())
+
```
Comment (dzhulgakov) [inline torch/serialization.py]: can you remove unrelated whitespace changes in this file?

File: test/test_torch.py [hunk 3]
```diff
@@ -3099,14 +3099,16 @@ def test_qtensor_permute(self):
     def test_qtensor_load_save(self):
         scale = 2.0
         zero_point = 10
-        r = torch.ones(15, dtype=torch.float) * 2
+        r = np.random.rand(100, 30) * 2 - 4
```
Comment (dzhulgakov) [inline test/test_torch.py]: nit: why not torch.rand()

File: torch/nn/quantized/modules/linear.py (hunk 1 above)
Comment (jerryzh168) [reply torch/nn/quantized/modules/linear.py]: @dzhulgakov sure, I'll change after that is landed.

**Review by dzhulgakov — APPROVED** [review]
#21933 is landing, so maybe you can rebase

File: torch/nn/quantized/modules/linear.py [hunk 4]
```diff
@@ -57,3 +66,35 @@ def from_float(mod):
             qbias = torch.quantize_linear(
                 bias, output_scale, output_zero_point, torch.qint32)
         return Linear(qweight, qbias, output_scale, output_zero_point)
+
+    # TODO: remove after https://github.com/pytorch/pytorch/pull/21933 is landed
```
Comment (dzhulgakov) [inline torch/nn/quantized/modules/linear.py]: you can remove it now

File: test/test_nn_quantized.py [hunk 5]
```diff
@@ -47,6 +47,43 @@ def test_linear_api(self):
...
+        model_dict = qLinear.state_dict()
+        self.assertEqual(model_dict['weight'], W_q)
+        with tempfile.NamedTemporaryFile() as f:
+            torch.save(model_dict, f)
+            f.seek(0)
+            loaded_dict = torch.load(f)
+        for key in model_dict:
+            self.assertEqual(model_dict[key], loaded_dict[key])
+        # TODO: should we do this?
+        qLinear2 = nnq.Linear.from_float(torch.nn.Linear(output_channels, input_channels))
```
Comment (jerryzh168) [inline test/test_nn_quantized.py]: @dzhulgakov, here we need to construct a dummy nnq.Linear module, how should we do that?

File: torch/nn/quantized/modules/linear.py [hunk 6]
```diff
@@ -61,3 +61,17 @@ def forward(self, x):
...
+
+    def _save_to_state_dict(self, destination, prefix, keep_vars):
+        super()._save_to_state_dict(destination, prefix, keep_vars)
+        destination[prefix + 'weight'] = torch.ops.quantized.fbgemm_linear_unpack(destination[prefix + '_packed_weight'])
+        destination.pop(prefix + '_packed_weight')
+
+    def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+                              missing_keys, unexpected_keys, error_msgs):
+        self._packed_weight = torch.ops.quantized.fbgemm_linear_prepack(state_dict[prefix + 'weight'])
+        self.bias = state_dict[prefix + 'bias']
```
Comment (dzhulgakov) [inline torch/nn/quantized/modules/linear.py]: it's better to preserve the identity of the bias buffer, so do `self.bias.copy_(state_dict[prefix + 'bias'])`?

File: torch/nn/quantized/modules/linear.py [hunk 7]
```diff
@@ -61,3 +61,17 @@ def forward(self, x):
...
+        state_dict.pop(prefix + 'weight')
+        super()._load_from_state_dict(state_dict, prefix, local_metadata, False,
```
Comment (dzhulgakov) [inline torch/nn/quantized/modules/linear.py]: given that the state_dict for linear is literally bias+weight, I don't think it makes sense to call parent method

File: torch/nn/quantized/modules/linear.py [hunk 8]
```diff
@@ -61,3 +61,17 @@ def forward(self, x):
             self.out_scale,
             self.out_zero_point)
         return Y_q
+
+    def _save_to_state_dict(self, destination, prefix, keep_vars):
+        super()._save_to_state_dict(destination, prefix, keep_vars)
```
Comment (dzhulgakov) [inline torch/nn/quantized/modules/linear.py]: given that the state_dict for linear is literally bias+weight, I don't think it makes sense to call parent method - you can just add 'bias' handling here

File: torch/nn/quantized/modules/linear.py (hunk 7 above)
Comment (jerryzh168) [reply torch/nn/quantized/modules/linear.py]: I can remove bias as well from state_dict. I saw there is a handling of hook in parent function, should that be called?

Comment (jerryzh168) [reply torch/nn/quantized/modules/linear.py]: there is 'output_scale' and 'output_zero_point' as well

File: torch/nn/quantized/modules/linear.py (hunk 8 above)
Comment (jerryzh168) [reply torch/nn/quantized/modules/linear.py]: there are 'output_scale' and 'output_zero_point' as well
